src/main/kotlin/com/atlassian/performance/tools/awsinfrastructure/api/AwsDatasetModification.kt

```diff
@@ -1,77 +1,59 @@
-package com.atlassian.performance.tools.awsinfrastructure
+package com.atlassian.performance.tools.awsinfrastructure.api
 
 import com.atlassian.performance.tools.aws.api.Aws
 import com.atlassian.performance.tools.aws.api.Investment
 import com.atlassian.performance.tools.aws.api.Resource
-import com.atlassian.performance.tools.awsinfrastructure.api.CustomDatasetSource
-import com.atlassian.performance.tools.awsinfrastructure.api.Infrastructure
-import com.atlassian.performance.tools.awsinfrastructure.api.InfrastructureFormula
-import com.atlassian.performance.tools.awsinfrastructure.api.ProvisionedInfrastructure
 import com.atlassian.performance.tools.awsinfrastructure.api.hardware.C5NineExtraLargeEphemeral
 import com.atlassian.performance.tools.awsinfrastructure.api.jira.Jira
 import com.atlassian.performance.tools.awsinfrastructure.api.jira.StandaloneFormula
 import com.atlassian.performance.tools.awsinfrastructure.api.virtualusers.AbsentVirtualUsersFormula
 import com.atlassian.performance.tools.infrastructure.api.dataset.Dataset
 import com.atlassian.performance.tools.infrastructure.api.distribution.PublicJiraSoftwareDistribution
+import com.atlassian.performance.tools.workspace.api.RootWorkspace
 import com.atlassian.performance.tools.workspace.api.TestWorkspace
 import org.apache.logging.log4j.LogManager.getLogger
 import java.time.Duration
+import java.util.*
 import java.util.concurrent.TimeUnit
+import java.util.function.Consumer
+import java.util.function.Function
 
-internal class AwsDataset(
-    private val dataset: Dataset
-) {
+class AwsDatasetModification private constructor(
+    private val aws: Aws,
+    private val dataset: Dataset,
+    private val workspace: TestWorkspace,
+    private val newDatasetName: String,
+    private val formula: InfrastructureFormula<*>
+): Function<Consumer<Infrastructure<*>>, Dataset> {
 
-    fun modify(
-        aws: Aws,
-        workspace: TestWorkspace,
-        newDatasetName: String,
-        modification: (Infrastructure<*>) -> Unit
-    ): Dataset {
-        val provisionedInfrastructure = provision(aws, workspace)
+    override fun apply(modification: Consumer<Infrastructure<*>>): Dataset {
```

Comment:
@wyrzyk @dagguh @drspartez do you think `modification: Consumer<Infrastructure<*>>` should be a parameter to the constructor instead (with default no-op implementation in Builder)?

Reply:
1. No-op seems not to be a good default. 
2. We started to avoid defaults in a constructor in favour of builders. It'll be hard/impossible to consume the default value from java.